File: app/src/main/java/igrek/todotree/dagger/FactoryModule.java
package igrek.todotree.dagger;

import android.app.Activity;
import android.content.Context;
import android.support.v7.app.AppCompatActivity;

import javax.inject.Singleton;

import dagger.Lazy;
import dagger.Module;
import dagger.Provides;
import igrek.todotree.app.AppControllerService;
import igrek.todotree.app.AppData;
import igrek.todotree.logger.Logger;
import igrek.todotree.logger.LoggerFactory;
import igrek.todotree.service.access.AccessLogService;
import igrek.todotree.service.access.DatabaseLock;
import igrek.todotree.service.access.QuickAddService;
import igrek.todotree.service.backup.BackupManager;
import igrek.todotree.service.clipboard.SystemClipboardManager;
import igrek.todotree.service.clipboard.TreeClipboardManager;
import igrek.todotree.service.commander.SecretCommander;
import igrek.todotree.service.filesystem.ExternalCardService;
import igrek.todotree.service.filesystem.FilesystemService;
import igrek.todotree.service.history.ChangesHistory;
import igrek.todotree.service.history.LinkHistoryService;
import igrek.todotree.service.preferences.Preferences;
import igrek.todotree.service.resources.UserInfoService;
import igrek.todotree.service.statistics.StatisticsLogService;
import igrek.todotree.service.summary.AlarmService;
import igrek.todotree.service.summary.DailySummaryService;
import igrek.todotree.service.summary.NotificationService;
import igrek.todotree.service.system.SoftKeyboardService;
import igrek.todotree.service.tree.ContentTrimmer;
import igrek.todotree.service.tree.TreeManager;
import igrek.todotree.service.tree.TreeMover;
import igrek.todotree.service.tree.TreeScrollCache;
import igrek.todotree.service.tree.TreeSelectionManager;
import igrek.todotree.service.tree.persistence.TreePersistenceService;
import igrek.todotree.ui.GUI;

/**
 * Module with providers. These classes can be injected
 */
@Module
public class FactoryModule {
	
	private AppCompatActivity activity;
	
	public FactoryModule(AppCompatActivity activity) {
		this.activity = activity;
	}
	
	@Provides
	protected AppCompatActivity provideAppCompatActivity() {
		return activity;
	}
	
	@Provides
	protected Activity provideActivity() {
		return activity;
	}
	
	@Provides
	protected Context provideContext() {
		return activity;
	}
	
	@Provides
	protected Logger provideLogger() {
		return LoggerFactory.getLogger();
	}
	
	@Provides
	@Singleton
	protected AppControllerService provideAppControllerSerivce(Activity activity) {
		return new AppControllerService(activity);
	}
	
	@Provides
	@Singleton
	protected FilesystemService provideFilesystemService(Logger logger, Activity activity, ExternalCardService externalCardService) {
		return new FilesystemService(logger, activity, externalCardService);
	}
	
	@Provides
	@Singleton
	protected TreeManager provideTreeManager(ChangesHistory changesHistory) {
		return new TreeManager(changesHistory);
	}
	
	@Provides
	@Singleton
	protected Preferences providePreferences(Context context, Logger logger) {
		return new Preferences(context, logger);
	}
	
	@Provides
	@Singleton
	protected BackupManager provideBackupManager(Preferences preferences, FilesystemService filesystem, Logger logger) {
		return new BackupManager(preferences, filesystem, logger);
	}
	
	@Provides
	@Singleton
	protected UserInfoService provideUserInfoService(Activity activity, GUI gui, Logger logger) {
		return new UserInfoService(activity, gui, logger);
	}
	
	@Provides
	@Singleton
	protected TreePersistenceService providePersistenceService() {
		return new TreePersistenceService();
	}
	
	@Provides
	@Singleton
	protected GUI provideGUI(AppCompatActivity activity) {
		return new GUI(activity);
	}
	
	@Provides
	@Singleton
	protected SystemClipboardManager provideSystemClipboardManager(Activity activity) {
		return new SystemClipboardManager(activity);
	}
	
	@Provides
	@Singleton
	protected AppData provideAppData() {
		return new AppData();
	}
	
	@Provides
	@Singleton
	protected DatabaseLock provideDatabaseLock(Preferences preferences, Logger logger, AccessLogService accessLogService) {
		return new DatabaseLock(preferences, logger, accessLogService);
	}
	
	@Provides
	@Singleton
	protected AccessLogService provideAccessLogService(FilesystemService filesystem, Preferences preferences, Logger logger) {
		return new AccessLogService(filesystem, preferences, logger);
	}
	
	@Provides
	@Singleton
	protected ChangesHistory provideChangesHistory() {
		return new ChangesHistory();
	}
	
	@Provides
	@Singleton
	protected ContentTrimmer provideContentTrimmer() {
		return new ContentTrimmer();
	}
	
	@Provides
	@Singleton
	protected TreeScrollCache provideTreeScrollCache() {
		return new TreeScrollCache();
	}
	
	@Provides
	@Singleton
	protected TreeClipboardManager provideTreeClipboardManager() {
		return new TreeClipboardManager();
	}
	
	@Provides
	@Singleton
	protected TreeSelectionManager provideTreeSelectionManager() {
		return new TreeSelectionManager();
	}
	
	@Provides
	@Singleton
	protected TreeMover provideTreeMover() {
		return new TreeMover();
	}
	
	@Provides
	@Singleton
	protected SecretCommander provideCommander(Logger logger, Preferences preferences, UserInfoService userInfo) {
		return new SecretCommander(logger, preferences, userInfo);
	}
	
	@Provides
	@Singleton
	protected StatisticsLogService provideStatisticsLogService(FilesystemService filesystem, Preferences preferences) {
		return new StatisticsLogService(filesystem, preferences);
	}
	
	@Provides
	@Singleton
	protected ExternalCardService provideExternalCardService(Logger logger) {
		return new ExternalCardService(logger);
	}
	
	@Provides
	@Singleton
	protected QuickAddService provideQuickAddService(Logger logger, Activity activity, TreeManager treeManager, SoftKeyboardService softKeyboardService, Lazy<GUI> gui) {
		return new QuickAddService(logger, activity, treeManager, softKeyboardService, gui);
	}
	
	@Provides
	@Singleton
	protected LinkHistoryService provideLinkHistoryService() {
		return new LinkHistoryService();
	}
	
	@Provides
	@Singleton
	protected NotificationService provideNotificationService() {
		return new NotificationService();
	}
	
	@Provides
	@Singleton
	protected AlarmService provideAlarmService(Activity activity) {
		return new AlarmService(activity);
	}
	
	@Provides
	@Singleton
	protected SoftKeyboardService provideSoftKeyboardService(Activity activity) {
		return new SoftKeyboardService(activity);
	}
	
	@Provides
	@Singleton
	protected DailySummaryService provideDailySummaryService(Activity activity, NotificationService notificationService, StatisticsLogService statisticsLogService) {
		return new DailySummaryService(activity, notificationService, statisticsLogService);
	}
	
}

File: app/src/main/java/igrek/todotree/dagger/FactoryModule.kt
package igrek.todotree.dagger


import android.app.Activity
import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import dagger.Lazy
import dagger.Module
import dagger.Provides
import igrek.todotree.activity.ActivityController
import igrek.todotree.activity.AppInitializer
import igrek.todotree.activity.OptionSelectDispatcher
import igrek.todotree.app.AppData
import igrek.todotree.info.logger.Logger
import igrek.todotree.info.logger.LoggerFactory
import igrek.todotree.service.access.AccessLogService
import igrek.todotree.service.access.DatabaseLock
import igrek.todotree.service.access.QuickAddService
import igrek.todotree.service.backup.BackupManager
import igrek.todotree.service.clipboard.SystemClipboardManager
import igrek.todotree.service.clipboard.TreeClipboardManager
import igrek.todotree.service.commander.SecretCommander
import igrek.todotree.service.filesystem.FilesystemService
import igrek.todotree.service.history.ChangesHistory
import igrek.todotree.service.history.LinkHistoryService
import igrek.todotree.service.preferences.Preferences
import igrek.todotree.service.remote.RemoteDbRequester
import igrek.todotree.service.remote.RemotePushService
import igrek.todotree.service.resources.UserInfoService
import igrek.todotree.service.statistics.StatisticsLogService
import igrek.todotree.service.summary.AlarmService
import igrek.todotree.service.summary.DailySummaryService
import igrek.todotree.service.summary.NotificationService
import igrek.todotree.service.system.SoftKeyboardService
import igrek.todotree.service.tree.*
import igrek.todotree.service.tree.persistence.TreePersistenceService
import igrek.todotree.system.PackageInfoService
import igrek.todotree.system.PermissionService
import igrek.todotree.system.SystemKeyDispatcher
import igrek.todotree.system.WindowManagerService
import igrek.todotree.system.filesystem.ExternalCardService
import igrek.todotree.ui.ExplosionService
import igrek.todotree.ui.GUI
import okhttp3.OkHttpClient
import javax.inject.Singleton

/**
 * Module with providers. These classes can be injected
 */
@Module
open class FactoryModule(private val activity: AppCompatActivity) {

    @Provides
    open fun provideContext(): Context {
        return activity.applicationContext
    }

    @Provides
    fun provideActivity(): Activity {
        return activity
    }

    @Provides
    open fun provideAppCompatActivity(): AppCompatActivity {
        return activity
    }

    @Provides
    open fun provideLogger(): Logger {
        return LoggerFactory.logger
    }

    /* Services */

    @Provides
    @Singleton
    fun aActivityController(): ActivityController = ActivityController()

    @Provides
    @Singleton
    fun aAppInitializer(): AppInitializer = AppInitializer()

    @Provides
    @Singleton
    fun aOptionSelectDispatcher(): OptionSelectDispatcher = OptionSelectDispatcher()

    @Provides
    @Singleton
    fun aSystemKeyDispatcher(): SystemKeyDispatcher = SystemKeyDispatcher()

    @Provides
    @Singleton
    fun aScreenService(): WindowManagerService = WindowManagerService()

    @Provides
    @Singleton
    fun aSoftKeyboardService(activity: Activity): SoftKeyboardService = SoftKeyboardService(activity)


    @Provides
    @Singleton
    fun aPermissionService(): PermissionService = PermissionService()

    @Provides
    @Singleton
    fun aPackageInfoService(): PackageInfoService = PackageInfoService()

    @Provides
    @Singleton
    fun aOkHttpClient(): OkHttpClient = OkHttpClient()

    @Provides
    @Singleton
    open fun aFilesystemService(context: Context): FilesystemService = FilesystemService(context)

    @Provides
    @Singleton
    fun aTreeManager(changesHistory: ChangesHistory): TreeManager = TreeManager(changesHistory)

    @Provides
    @Singleton
    open fun aPreferences(context: Context): Preferences = Preferences(context)

    @Provides
    @Singleton
    fun aBackupManager(filesystem: FilesystemService): BackupManager = BackupManager(filesystem)

    @Provides
    @Singleton
    open fun aUserInfoService(activity: Activity, gui: GUI): UserInfoService = UserInfoService(activity, gui)

    @Provides
    @Singleton
    fun aTreePersistenceService(): TreePersistenceService = TreePersistenceService()

    @Provides
    @Singleton
    open fun aGUI(activity: AppCompatActivity): GUI = GUI(activity)

    @Provides
    @Singleton
    open fun aSystemClipboardManager(activity: Activity): SystemClipboardManager = SystemClipboardManager(activity)

    @Provides
    @Singleton
    fun aAppData(): AppData = AppData()

    @Provides
    @Singleton
    open fun aDatabaseLock(preferences: Preferences, accessLogService: AccessLogService): DatabaseLock = DatabaseLock(preferences, accessLogService)

    @Provides
    @Singleton
    fun aAccessLogService(filesystem: FilesystemService): AccessLogService = AccessLogService(filesystem)

    @Provides
    @Singleton
    fun aChangesHistory(): ChangesHistory = ChangesHistory()

    @Provides
    @Singleton
    fun aContentTrimmer(): ContentTrimmer = ContentTrimmer()

    @Provides
    @Singleton
    fun aTreeScrollCache(): TreeScrollCache = TreeScrollCache()

    @Provides
    @Singleton
    fun aTreeClipboardManager(): TreeClipboardManager = TreeClipboardManager()

    @Provides
    @Singleton
    fun aTreeSelectionManager(): TreeSelectionManager = TreeSelectionManager()

    @Provides
    @Singleton
    fun aTreeMover(): TreeMover = TreeMover()

    @Provides
    @Singleton
    fun aSecretCommander(preferences: Preferences, userInfo: UserInfoService): SecretCommander = SecretCommander(preferences, userInfo)

    @Provides
    @Singleton
    fun aStatisticsLogService(filesystem: FilesystemService): StatisticsLogService = StatisticsLogService(filesystem)

    @Provides
    @Singleton
    open fun aExternalCardService(): ExternalCardService = ExternalCardService()

    @Provides
    @Singleton
    fun aQuickAddService(activity: Activity, treeManager: TreeManager, gui: dagger.Lazy<GUI>, userInfoService: UserInfoService, activityController: ActivityController): QuickAddService = QuickAddService(activity, treeManager, gui, userInfoService, activityController)

    @Provides
    @Singleton
    fun aLinkHistoryService(): LinkHistoryService = LinkHistoryService()

    @Provides
    @Singleton
    fun aNotificationService(): NotificationService = NotificationService()

    @Provides
    @Singleton
    fun aAlarmService(activity: Activity): AlarmService = AlarmService(activity)

    @Provides
    @Singleton
    fun aDailySummaryService(activity: Activity, notificationService: NotificationService, statisticsLogService: StatisticsLogService): DailySummaryService = DailySummaryService(activity, notificationService, statisticsLogService)

    @Provides
    @Singleton
    fun aRemoteDbRequester(preferences: Preferences, activity: Activity): RemoteDbRequester = RemoteDbRequester(preferences, activity)

    @Provides
    @Singleton
    fun aRemotePushService(activity: Activity, treeManager: TreeManager, gui: Lazy<GUI>, userInfoService: UserInfoService, remoteDbRequester: RemoteDbRequester): RemotePushService = RemotePushService(activity, treeManager, gui, userInfoService, remoteDbRequester)

    @Provides
    @Singleton
    fun aExplosionService(activity: Activity): ExplosionService = ExplosionService(activity)

    /*
	 * Empty service pattern:
	@Provides
    @Singleton
    fun a():  = ()

	 */

}
